Add K3s Hardening Guide and Self Assessment (#2882)
rancher/docs: content/k3s/latest/en/security/self_assessment/_index.md

---
title: "CIS Self Assessment Guide"
weight: 90
---

> ### CIS Kubernetes Benchmark v1.5 - K3s v1.17, v1.18, & v1.19

Reviewer: Are we allowed to redistribute this? If so, are there any copyright or attribution guidelines we need to follow?

Author: We checked it out when we did RKE2. Seemed legit.

#### Overview

> This document is a companion to the K3s security hardening guide. The hardening guide provides prescriptive guidance for hardening a production installation of K3s, and this benchmark guide is meant to help you evaluate the level of security of the hardened cluster against each control in the CIS Kubernetes benchmark. It is to be used by K3s operators, security teams, auditors, and decision makers.

Reviewer 2: Nit - utilize a hyphen

```suggestion
This document is a companion to the K3s security hardening guide. The hardening guide provides prescriptive guidance for hardening a production installation of K3s, and this benchmark guide is meant to help you evaluate the level of security of the hardened cluster against each control in the CIS Kubernetes benchmark. It is to be used by K3s operators, security teams, auditors, and decision-makers.
```

Author: Updated.

```bash
journalctl -u k3s | grep "Running kube-apiserver" | tail -n1 | grep "token-auth-file"
```

Verify that the `--token-auth-file` argument does not exist.

**Remediation:**
By default, K3s does not run with basic authentication enabled. No manual remediation is needed.

#### 1.2.4
Ensure that the `--kubelet-https` argument is set to true (Scored)

<details>
<summary>Rationale</summary>
Connections from apiserver to kubelets could potentially carry sensitive data such as secrets and keys. It is thus important to use in-transit encryption for any communication between the apiserver and kubelets.
</details>

**Result:** Not Applicable

**Audit:**
Run the below command on the master node.

```bash
journalctl -u k3s | grep "Running kube-apiserver" | tail -n1 | grep "kubelet-https"
```

Verify that the `--kubelet-https` argument does not exist.

**Remediation:**
By default, K3s kube-apiserver doesn't run with the `--kubelet-https` parameter as it runs with TLS. No manual remediation is needed.

#### 1.2.5
Ensure that the `--kubelet-client-certificate` and `--kubelet-client-key` arguments are set as appropriate (Scored)

<details>
<summary>Rationale</summary>

> The apiserver, by default, does not authenticate itself to the kubelet's HTTPS endpoints. The requests from the apiserver are treated anonymously. You should set up certificate- based kubelet authentication to ensure that the apiserver authenticates itself to kubelets when submitting requests.

Reviewer 2: Remove extra space

```suggestion
The apiserver, by default, does not authenticate itself to the kubelet's HTTPS endpoints. The requests from the apiserver are treated anonymously. You should set up certificate-based kubelet authentication to ensure that the apiserver authenticates itself to kubelets when submitting requests.
```

Author: Updated.

K3s would need to have the `PodSecurityPolicy` admission plugin enabled by passing it as an argument to K3s. `--kube-apiserver-arg='enable-admission-plugins=PodSecurityPolicy`.

#### 1.2.17
Ensure that the admission control plugin `NodeRestriction` is set (Scored)
<details>
<summary>Rationale</summary>
Using the `NodeRestriction` plug-in ensures that the kubelet is restricted to the `Node` and `Pod` objects that it could modify as defined. Such kubelets will only be allowed to modify their own `Node` API object, and only modify `Pod` API objects that are bound to their node.

</details>

**Result:** Pass

**Audit:**
Run the below command on the master node.

```bash
journalctl -u k3s | grep "Running kube-apiserver" | tail -n1 | grep "enable-admission-plugins"
```

Verify that the `--enable-admission-plugins` argument is set to a value that includes `NodeRestriction`.

**Remediation:**
K3s would need to have the `NodeRestriction` admission plugin enabled by passing it as an argument to K3s. `--kube-apiserver-arg='enable-admission-plugins=NodeRestriction`.

#### 1.2.18
Ensure that the `--insecure-bind-address` argument is not set (Scored)
<details>
<summary>Rationale</summary>

> If you bind the apiserver to an insecure address, basically anyone who could connect to it over the insecure port, would have unauthenticated and unencrypted access to your master node. The apiserver doesn't do any authentication checking for insecure binds and traffic to the Insecure API port is not encrpyted, allowing attackers to potentially read sensitive data in transit.

Reviewer 2: Spelling 

```suggestion
If you bind the apiserver to an insecure address, basically anyone who could connect to it over the insecure port, would have unauthenticated and unencrypted access to your master node. The apiserver doesn't do any authentication checking for insecure binds and traffic to the Insecure API port is not encrypted, allowing attackers to potentially read sensitive data in transit.
```

#### 1.2.28
Ensure that the `--service-account-key-file` argument is set as appropriate (Scored)
<details>
<summary>Rationale</summary>
By default, if no `--service-account-key-file` is specified to the apiserver, it uses the private key from the TLS serving certificate to verify service account tokens. To ensure that the keys for service account tokens could be rotated as needed, a separate public/private key pair should be used for signing service account tokens. Hence, the public key should be specified to the apiserver with `--service-account-key-file`.
</details>

**Result:** Pass

**Audit:**
Run the below command on the master node.

```bash
journalctl -u k3s | grep "Running kube-apiserver" | tail -n1 | grep "service-account-key-file"
```

Verify that the `--service-account-key-file` argument exists and is set as appropriate.

**Remediation:**
By default, K3s sets the `--service-account-key-file` explicitly. No manual remediation needed.

#### 1.2.29
Ensure that the `--etcd-certfile` and `--etcd-keyfile` arguments are set as appropriate (Scored)
<details>
<summary>Rationale</summary>

> etcd is a highly-available key value store used by Kubernetes deployments for persistent storage of all of its REST API objects. These objects are sensitive in nature and should be protected by client authentication. This requires the API server to identify itself to the etcd server using a client certificate and key.

Reviewer 2: Utilize a hyphen

```suggestion
etcd is a highly-available key-value store used by Kubernetes deployments for persistent storage of all of its REST API objects. These objects are sensitive in nature and should be protected by client authentication. This requires the API server to identify itself to the etcd server using a client certificate and key.
```

Reviewer 2: I see one or more instances of this as well ("key value" should be "key-value")

Author: Updated all.

#### 1.2.34
Ensure that encryption providers are appropriately configured (Scored)
<details>
<summary>Rationale</summary>
Where `etcd` encryption is used, it is important to ensure that the appropriate set of encryption providers is used. Currently, the `aescbc`, `kms` and `secretbox` are likely to be appropriate options.
</details>

**Result:** Pass

**Remediation:**
Follow the Kubernetes documentation and configure a `EncryptionConfig` file.
In this file, choose **aescbc**, **kms** or **secretbox** as the encryption provider.

**Audit:**
Run the below command on the master node.

```bash
grep aescbc /path/to/encryption-config.json
```

Run the below command on the master node.

Verify that aescbc is set as the encryption provider for all the desired resources.

**Remediation**
K3s server needs to be run with the following, `--secrets-encryption=true`, and verify that one of the allowed encryption providers is present.

#### 1.2.35
Ensure that the API Server only makes use of Strong Cryptographic Ciphers (Not Scored)

<details>
<summary>Rationale</summary>

> TLS ciphers have had a number of known vulnerabilities and weaknesses, which can reduce the protection provided by them. By default Kubernetes supports a number of TLS ciphersuites including some that have security concerns, weakening the protection provided.

Reviewer 2: Uncertain - should "ciphersuites" be "cipher suites" ?

```suggestion
TLS ciphers have had a number of known vulnerabilities and weaknesses, which can reduce the protection provided by them. By default Kubernetes supports a number of TLS cipher suites including some that have security concerns, weakening the protection provided.
```

Author: Updated.

Reviewer 2: I still see one match at https://github.com/rancher/docs/pull/2882/files#diff-bfe898ebe76c182dc5b76cf3cc88d6d8014de491209af72c97cd16b06dfdd3acR1956 but not a big deal.

#### 1.3.5
Ensure that the `--root-ca-file` argument is set as appropriate (Scored)
<details>
<summary>Rationale</summary>
Processes running within pods that need to contact the API server must verify the API server's serving certificate. Failing to do so could be a subject to man-in-the-middle attacks.

Providing the root certificate for the API server's serving certificate to the controller manager with the `--root-ca-file` argument allows the controller manager to inject the trusted bundle into pods so that they can verify TLS connections to the API server.
</details>

**Result:** Pass

**Audit:**
Run the below command on the master node.

```bash
journalctl -u k3s | grep "Running kube-controller-manager" | tail -n1 | grep "root-ca-file"
```

Verify that the `--root-ca-file` argument exists and is set to a certificate bundle file containing the root certificate for the API server's serving certificate

**Remediation:**
By default, K3s sets the `--root-ca-file` argument with the root ca file. No manual remediation needed.

#### 1.3.6
Ensure that the `RotateKubeletServerCertificate` argument is set to `true` (Scored)
<details>
<summary>Rationale</summary>

> `RotateKubeletServerCertificate` causes the kubelet to both request a serving certificate after bootstrapping its client credentials and rotate the certificate as its existing credentials expire. This automated periodic rotation ensures that the there are no downtimes due to expired certificates and thus addressing availability in the CIA security triad.

Reviewer 2: Remove erroneous "the"

```suggestion
`RotateKubeletServerCertificate` causes the kubelet to both request a serving certificate after bootstrapping its client credentials and rotate the certificate as its existing credentials expire. This automated periodic rotation ensures that there are no downtimes due to expired certificates and thus addressing availability in the CIA security triad.
```

Reviewer 2: There are a few other areas where we have "the there"

Author: Updated.

Reviewer 2: I still see several "the there" entires like https://github.com/rancher/docs/pull/2882/files#diff-bfe898ebe76c182dc5b76cf3cc88d6d8014de491209af72c97cd16b06dfdd3acR1924 and https://github.com/rancher/docs/pull/2882/files#diff-bfe898ebe76c182dc5b76cf3cc88d6d8014de491209af72c97cd16b06dfdd3acR1941

Note: This recommendation only applies if you let kubelets get their certificates from the API server. In case your kubelet certificates come from an outside authority/tool (e.g. Vault) then you need to take care of rotation yourself.
</details>

**Result:** Not Applicable

**Audit:**
Run the below command on the master node.

```bash
journalctl -u k3s | grep "Running kube-controller-manager" | tail -n1 | grep "RotateKubeletServerCertificate"
```

Verify that RotateKubeletServerCertificateargument exists and is set to true.

**Remediation:**

> By default, K3s implements it's own logic for certificate generation and rotation.

Reviewer 2: remove apostrophe 

```suggestion
By default, K3s implements its own logic for certificate generation and rotation.
```

Reviewer 2: There are at least a couple other places that also have this issue.

Author: Updated.

Reviewer 2: I still see https://github.com/rancher/docs/pull/2882/files#diff-bfe898ebe76c182dc5b76cf3cc88d6d8014de491209af72c97cd16b06dfdd3acR1934 and https://github.com/rancher/docs/pull/2882/files#diff-bfe898ebe76c182dc5b76cf3cc88d6d8014de491209af72c97cd16b06dfdd3acR1949

**Remediation:**
By default, K3s uses a config file for etcd that can be found at `/var/lib/rancher/k3s/server/db/etcd/config` and the `trusted-ca-file` parameters in it are set to unique values specific to etcd. No manual remediation needed.

## 3 Control Plane Configuration

### 3.1 Authentication and Authorization

#### 3.1.1
Client certificate authentication should not be used for users (Not Scored)
<details>
<summary>Rationale</summary>
With any authentication mechanism the ability to revoke credentials if they are compromised or no longer required, is a key control. Kubernetes client certificate authentication does not allow for this due to a lack of support for certificate revocation.
</details>

**Result:** Not Scored - Operator Dependent

**Audit:**
Review user access to the cluster and ensure that users are not making use of Kubernetes client certificate authentication.

**Remediation:**
Alternative mechanisms provided by Kubernetes such as the use of OIDC should be implemented in place of client certificates.

### 3.2 Logging

#### 3.2.1
Ensure that a minimal audit policy is created (Scored)
<details>
<summary>Rationale</summary>

> Logging is an important detective control for all systems, to detect potential unauthorised access.

Reviewer 2: Spelling

```suggestion
Logging is an important detective control for all systems, to detect potential unauthorized access.
```

Author: Updated.

</details>

> **Result:** Does not pass. See the [Hardening Guide](../hardening_guide/_index.md) for details.

Reviewer 2: Throughout, just replace _index.md with null. I see nine entries with this and this won't work with Hugo.

**Audit:**
Run the below command on the master node.

```bash
journalctl -u k3s | grep "Running kube-apiserver" | tail -n1 | grep "audit-policy-file"
```

> Verify that the `--audit-policy-file` is set. Review the contents of the file specified and ensure that it contains avalid audit policy.

Reviewer 2: Add space

```suggestion
Verify that the `--audit-policy-file` is set. Review the contents of the file specified and ensure that it contains a valid audit policy.
```

Author: Updated.

</details>

**Result:** Not Applicable

**Audit:**
Run the below command on the master node.

```bash
stat -c %U:%G /var/lib/rancher/k3s/agent/kubeproxy.kubeconfig
root:root
```

Verify that if a file is specified and it exists, the permissions are 644 or more restrictive.

**Remediation:**
K3s runs `kube-proxy` in process and does not use a config file.

#### 4.1.5
Ensure that the kubelet.conf file permissions are set to `644` or more restrictive (Scored)
<details>
<summary>Rationale</summary>
The `kubelet.conf` file is the kubeconfig file for the node, and controls various parameters that set the behavior and identity of the worker node. You should restrict its file permissions to maintain the integrity of the file. The file should be writable by only the administrators on the system.
</details>

**Result:** Pass

**Audit:**
Run the below command on the worker node.

```bash
stat -c %a /var/lib/rancher/k3s/agent/kubelet.kubeconfig
644
```

**Remediation:**

> By derfault, K3s creates `kubelet.kubeconfig` with `644` permissions. No manual remediation needed.

Reviewer 2: Spelling

```suggestion
By default, K3s creates `kubelet.kubeconfig` with `644` permissions. No manual remediation needed.
```

Author: Updated.

</details>

**Result:** Pass

**Audit:**
Run the below command on the master node.

```bash
journalctl -u k3s | grep "Running kubelet" | tail -n1 | grep -E 'tls-cert-file|tls-private-key-file'
```

Verify the `--tls-cert-file` and `--tls-private-key-file` arguments are present and set appropriately.

**Remediation:**
By default, K3s sets the `--tls-cert-file` and `--tls-private-key-file` arguments when executing the kubelet process.

#### 4.2.11
Ensure that the `--rotate-certificates` argument is not set to `false` (Scored)
<details>
<summary>Rationale</summary>

The `--rotate-certificates` setting causes the kubelet to rotate its client certificates by creating new CSRs as its existing credentials expire. This automated periodic rotation ensures that the there is no downtime due to expired certificates and thus addressing availability in the CIA security triad.

**Note:** This recommendation only applies if you let kubelets get their certificates from the API server. In case your kubelet certificates come from an outside authority/tool (e.g. Vault) then you need to take care of rotation yourself.

> **Note:**This feature also require the `RotateKubeletClientCertificate` feature gate to be enabled (which is the default since Kubernetes v1.7)

Reviewer 2: Spelling

```suggestion
**Note:**This feature also requires the `RotateKubeletClientCertificate` feature gate to be enabled (which is the default since Kubernetes v1.7)
```

Author: Updated.

Note: This recommendation only applies if you let kubelets get their certificates from the API server. In case your kubelet certificates come from an outside authority/tool (e.g. Vault) then you need to take care of rotation yourself.
</details>

**Result:** Not Applicable

**Remediation:**
By default, K3s implements it's own logic for certificate generation and rotation.

#### 4.2.13
Ensure that the Kubelet only makes use of Strong Cryptographic Ciphers (Not Scored)
<details>
<summary>Rationale</summary>
TLS ciphers have had a number of known vulnerabilities and weaknesses, which can reduce the protection provided by them. By default Kubernetes supports a number of TLS ciphersuites including some that have security concerns, weakening the protection provided.
</details>

**Result:** Not Scored - Operator Dependent

**Remediation:**

> Configuration of the parameter is dependent on your use case. Please see the CIS Kubernetes Benchmark for suggestions on configuring this for your usecase.

Reviewer 2: Add hyphen

```suggestion
Configuration of the parameter is dependent on your use case. Please see the CIS Kubernetes Benchmark for suggestions on configuring this for your use-case.
```

Author: Updated.

### 5.2 Pod Security Policies

#### 5.2.1
Minimize the admission of containers wishing to share the host process ID namespace (Scored)
<details>
<summary>Rationale</summary>
Privileged containers have access to all Linux Kernel capabilities and devices. A container running with full privileges can do almost everything that the host can do. This flag exists to allow special use-cases, like manipulating the network stack and accessing devices.

There should be at least one PodSecurityPolicy (PSP) defined which does not permit privileged containers.

If you need to run privileged containers, this should be defined in a separate PSP and you should carefully check RBAC controls to ensure that only limited service accounts and users are given permission to access that PSP.
</details>

**Result:** Pass

**Audit:**
Run the below command on the master node.

```bash
kubectl describe psp <psp_name> | grep MustRunAsNonRoot
```

Verify that the result is `Rule:  MustRunAsNonRoot`.

**Remediation:**

> An operator should apply a PodSecurityPolicy that sets the `Rule` value to `MustRunAsNonRoot`. An example of this can be found in the [Hardening Guide](../hardening_guilde/).

Reviewer 2: Spelling fix

```suggestion
An operator should apply a PodSecurityPolicy that sets the `Rule` value to `MustRunAsNonRoot`. An example of this can be found in the [Hardening Guide](../hardening_guide/).
```

> An operator should apply a PodSecurityPolicy that sets the `Rule` value to `MustRunAsNonRoot`. An example of this can be found in the [Hardening Guide](../hardening_guilde/_index.md).

Reviewer 2: Fix link

```suggestion
An operator should apply a PodSecurityPolicy that sets the `Rule` value to `MustRunAsNonRoot`. An example of this can be found in the [Hardening Guide](../hardening_guilde/).
```

Author: Updated.

#### 5.2.2
Minimize the admission of containers wishing to share the host process ID namespace (Scored)
<details>
<summary>Rationale</summary>
A container running in the host's PID namespace can inspect processes running outside the container. If the container also has access to ptrace capabilities this can be used to escalate privileges outside of the container.

There should be at least one PodSecurityPolicy (PSP) defined which does not permit containers to share the host PID namespace.

If you need to run containers which require hostPID, this should be defined in a separate PSP and you should carefully check RBAC controls to ensure that only limited service accounts and users are given permission to access that PSP.
</details>

**Result:** Pass

**Audit:**
Run the below command on the master node.

```bash
kubectl get psp -o json | jq .items[] | jq -r 'select((.spec.hostPID == null) or (.spec.hostPID == false))' | jq .metadata.name | wc -l | xargs -I {} echo '--count={}'
```

Verify that the returned count is 1.

**Remediation:**

> An operator should apply a PodSecurityPolicy that sets the `hostPID` value to false explicitly for the PSP it creates. An example of this can be found in the [Hardening Guide](../hardening_guide/_index.md).

Reviewer 2: Fix link

```suggestion
An operator should apply a PodSecurityPolicy that sets the `hostPID` value to false explicitly for the PSP it creates. An example of this can be found in the [Hardening Guide](../hardening_guide/).
```

There should be at least one PodSecurityPolicy (PSP) defined which does not permit containers to allow privilege escalation. The option exists (and is defaulted to true) to permit setuid binaries to run.

If you have need to run containers which use setuid binaries or require privilege escalation, this should be defined in a separate PSP and you should carefully check RBAC controls to ensure that only limited service accounts and users are given permission to access that PSP.
</details>

**Result:** Pass

**Audit:**
Run the below command on the master node.

```bash
kubectl get psp -o json | jq .items[] | jq -r 'select((.spec.allowPrivilegeEscalation == null) or (.spec.allowPrivilegeEscalation == false))' | jq .metadata.name | wc -l | xargs -I {} echo '--count={}'
```

Verify that the returned count is 1.

**Remediation:**
An operator should apply a PodSecurityPolicy that sets the `allowPrivilegeEscalation` value to false explicitly for the PSP it creates. An example of this can be found in the [Hardening Guide](../hardening_guide/_index.md).

#### 5.2.6
Minimize the admission of root containers (Not Scored)
<details>
<summary>Rationale</summary>

> Containers may run as any Linux user. Containers which run as the root user, whilst constrained by Container Runtime security features still have a escalated likelihood of container breakout.

Reviewer 2: Fix a/an usage

```suggestion
Containers may run as any Linux user. Containers which run as the root user, whilst constrained by Container Runtime security features still have an escalated likelihood of container breakout.
```

Author: Updated.

There should be at least one PodSecurityPolicy (PSP) defined which prevents containers with capabilities beyond the default set from launching.

If you need to run containers with additional capabilities, this should be defined in a separate PSP and you should carefully check RBAC controls to ensure that only limited service accounts and users are given permission to access that PSP.
</details>

**Result:** Not Scored

**Audit:**
Run the below command on the master node.

```bash
kubectl get psp
```

Verify that there are no PSPs present which have `allowedCapabilities` set to anything other than an empty array.

**Remediation:**
An operator should apply a PodSecurityPolicy that sets `allowedCapabilities` to anything other than an empty array. An example of this can be found in the [Hardening Guide](../hardening_guide/_index.md).

#### 5.2.9
Minimize the admission of containers with capabilities assigned (Not Scored)
<details>
<summary>Rationale</summary>
Containers run with a default set of capabilities as assigned by the Container Runtime. Capabilities are parts of the rights generally granted on a Linux system to the root user.

> In many cases applications running in containers do not require any capabilities to operate, so from the perspective of the principal of least privilege use of capabilities should be minimized.

Reviewer 2: Spelling

```suggestion
In many cases applications running in containers do not require any capabilities to operate, so from the perspective of the principle of least privilege use of capabilities should be minimized.
```

Author: Updated.

**Remediation:**
An operator should apply NetworkPolcyies that prevent unneeded traffic from traversing networks unnecessarily. An example of applying a NetworkPolcy can be found in the [Hardening Guide](../hardening_guide/_index.md).

### 5.4 Secrets Management

#### 5.4.1
Prefer using secrets as files over secrets as environment variables (Not Scored)
<details>
<summary>Rationale</summary>
It is reasonably common for application code to log out its environment (particularly in the event of an error). This will include any secret values passed in as environment variables, so secrets can easily be exposed to any user or entity who has access to the logs.
</details>

**Result:** Not Scored

**Audit:**
Run the following command to find references to objects which use environment variables defined from secrets.

```bash
kubectl get all -o jsonpath='{range .items[?(@..secretKeyRef)]} {.kind} {.metadata.name} {"\n"}{end}' -A
```

**Remediation:**
If possible, rewrite application code to read secrets from mounted secret files, rather than from environment variables.

#### 5.4.2
Consider external secret storage (Not Scored)
<details>
<summary>Rationale</summary>

> Kubernetes supports secrets as first-class objects, but care needs to be taken to ensure that access to secrets is carefully limited. Using an external secrets provider can ease the management of access to secrets, especially where secrests are used across both Kubernetes and non-Kubernetes environments.

Reviewer 2: Spelling

```suggestion
Kubernetes supports secrets as first-class objects, but care needs to be taken to ensure that access to secrets is carefully limited. Using an external secrets provider can ease the management of access to secrets, especially where secrets are used across both Kubernetes and non-Kubernetes environments.
```

Author: Updated.

</details>

**Result:** Not Scored

**Audit:**
Review your secrets management implementation.

**Remediation:**
Refer to the secrets management options offered by your cloud provider or a third-party secrets management solution.

### 5.5 Extensible Admission Control

#### 5.5.1
Configure Image Provenance using ImagePolicyWebhook admission controller (Not Scored)
<details>
<summary>Rationale</summary>
Kubernetes supports plugging in provenance rules to accept or reject the images in your deployments. You could configure such rules to ensure that only approved images are deployed in the cluster.
</details>

**Result:** Not Scored

**Audit:**
Review the pod definitions in your cluster and verify that image provenance is configured as appropriate.

**Remediation:**
Follow the Kubernetes documentation and setup image provenance.

### 5.6 Omitted

> The v1.5.1 guide skips 5.6 and goes from 5.5 to 5.7. We are including it here merely for explanation.

Reviewer 2: Nit: What guide? If we reference the official/upstream CIS guide should we link to it? I don't know what this guide is as a common user IMO.

```

Ensure that these namespaces are the ones you need and are adequately administered as per your requirements.

**Remediation:**
Follow the documentation and create namespaces for objects in your deployment as you need them.

#### 5.7.2
Ensure that the seccomp profile is set to `docker/default` in your pod definitions (Not Scored)
<details>
<summary>Rationale</summary>
Seccomp (secure computing mode) is used to restrict the set of system calls applications can make, allowing cluster administrators greater control over the security of workloads running in the cluster. Kubernetes disables seccomp profiles by default for historical reasons. You should enable it to ensure that the workloads have restricted actions available within the container.
</details>

**Result:** Not Scored

**Audit:**
Review the pod definitions in your cluster. It should create a line as below:

```yaml
annotations:
  seccomp.security.alpha.kubernetes.io/pod: docker/default
```

**Remediation:**
Review the Kubernetes documentation and if needed, apply a relevant PodSecurityPolicy.

#### 5.7.3
Apply Security Context to Your Pods and Containers (Not Scored)
<details>
<summary>Rationale</summary>

> A security context defines the operating system security settings (uid, gid, capabilities, SELinux role, etc..) applied to a container. When designing your containers and pods, make sure that you configure the security context for your pods, containers, and volumes. A security context is a property defined in the deployment yaml. It controls the security parameters that will be assigned to the pod/container/volume. There are two levels of security context: pod level security context, and container level security context.

Reviewer 2: Hyphen usage

```suggestion
A security context defines the operating system security settings (uid, gid, capabilities, SELinux role, etc..) applied to a container. When designing your containers and pods, make sure that you configure the security context for your pods, containers, and volumes. A security context is a property defined in the deployment yaml. It controls the security parameters that will be assigned to the pod/container/volume. There are two levels of security context: pod level security context, and container-level security context.
```

Author: Updated.

</details>

**Result:** Not Scored

**Audit:**
Review the pod definitions in your cluster and verify that you have security contexts defined as appropriate.

**Remediation:**
Follow the Kubernetes documentation and apply security contexts to your pods. For a suggested list of security contexts, you may refer to the CIS Security Benchmark.

#### 5.7.4
The default namespace should not be used (Scored)
<details>
<summary>Rationale</summary>
Resources in a Kubernetes cluster should be segregated by namespace, to allow for security controls to be applied at that level and to make it easier to manage resources.
</details>

**Result:** Pass

**Audit:**
Run the below command on the master node.

```bash
kubectl get all -n default
```

> The only entries there should be system managed resources such as the kubernetes service.

Reviewer 2: Hyphen usage

```suggestion
The only entries there should be system-managed resources such as the kubernetes service.
```

Author: Updated.
